OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs

```diff
@@ -0,0 +1,45 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class RemoveAttackFrontal : UpdateRule
+	{
+		public override string Name { get { return "Merge AttackFrontal into AttackFollow"; } }
+		public override string Description
+		{
+			get
+			{
+				return "The AttackFrontal trait is now deprecated and it's functionality incorporated\n" +
+				"into AttackFollow. The default AttackFrontal behaviour can be recreated by setting the\n" +
+				"FacingTolerance parameter to '0' and the new FireWhileMoving parameter to 'false'.";
+			}
+		}
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var attackNode = actorNode.LastChildMatching("AttackFrontal");
+			if (attackNode != null)
+			{
+				attackNode.RenameKey("AttackFollow");
+				var facingToleranceNode = attackNode.LastChildMatching("FacingTolerance");
+				if (facingToleranceNode == null)
+					attackNode.AddNode(new MiniYamlNode("FacingTolerance", "0"));
```

Comment:
We have a lot more ex-`AttackFrontal` actors than `AttackOmni` etc.

IMO it would make the code a lot more understandable if we could change the default value in `AttackBase` to 0 and likewise pushed all the `TargetInFiringArc`s into the base `CanAttack` implementation. The few actors that really do want to disable firing arcs can then specify `FacingTolerance: 128`.

Reply:
Every turreted actor we have currently has `FacingTolerance: 128`

Reply:
I consider that to be a large part of the problem - `FacingTolerance` should be used consistently for all actors, not used as a firing cone for frontal units and a turret allowance angle for turreted units.

Reply:
I suspect it would be easier to implement and review if we could fix this in a separate PR targeting bleed, then rebase this one on top.

Reply:
We currently don't have firing cones specifically for turrets; only for the bodies they are attached to. This is consistent for all actors. As it is we don't actually have proper turret allowance angles, and properly fixing that is something I consider something for a follow up. The aim of these first three PRs was to first apply some order to the jungle of attack traits we have before moving on to finetuning behaviours.

Reply:
> We currently don't have firing cones specifically for turrets; only for the bodies they are attached to.

Thanks for the correction, this is what I meant.

Trying to order the jungle up front then deal with the exceptions in followups didn't go very well during the activity overhaul, so i'd really like to avoid the same things repeating here.

Would we break any usecases by starting with a PR that changes turreted units to allow target acquisition from any angle and hooks up `FacingTolerance` as a firing cone?

Reply:
I don't know if it would break any usecases, but it would break my back.

I have zero motivation to change the whole plan after half a year of waiting. In fact, I'm starting to get burned out on this project as a whole, because of continued unreasonable demands like this. For what it's worth, I came very close to quitting after the whole move-undeploy debacle; I don't want to go through that bullshit again.

I really need to take a break from this project, but I would rather finish up any PRs that are still open in a tidy manner first. So please, accept this thing with minor corrections or close forever.

Reply:
I can't comment on individual change requests, but generally speaking, I can understand the getting burned out part. On average we're definitely taking far too long to give some PRs the attention they deserve, and that's pretty frustrating for everyone involved (fwiw, I'm in the same boat, I've had PRs sitting unreviewed for even longer in the past).

At least part of the problem for the lack of reviews in recent months, though, was that since release 20190314 had some show-stopperish issues (some of them outside our control, like Linux distros shipping broken Mono versions incompatible with OpenRA), we had to "hurry up" with getting another stable release out, which meant focusing on bugfixes over more refactorings.
Only that some of the larger refactorings we already merged turned that 'hurrying' into a ~5 month drag of almost nothing but bug-fixing, which - at least on my end, but I guess for everyone else involved, too - was absolute torture in terms of motivation.
tl;dr: The reason why I reduced my reviews to the bare minimum necessary to get the release out of the door over the last ~3-4 months was that I was, in fact, burned out, too. And I assume @pchote has been in a similar boat.

I'm afraid there isn't an easy near-term solution, other than taking a break when it gets too bad, since the core issue it boils down to is essentially lack of manpower.

We could consider (and have already discussed it in the past) shifting all official mods other than TS into maintenance mode (maybe keep them on bleed for testcases and potential future major updates, but otherwise freeze their official state in prep with only minor bugfix/maintenance/balance updates when sensible or necessary), so we could focus more on bleed in general and TS/next-gen mods in particular.

However, a) that increases the risk of accumulating too many subliminal bugs and other issues that only dedicated players would spot, making the following bugfix phase potentially even more devastating, and b) with the shroud perf improvement and @teinarss' HPA on the horizon, I feel that at least the current prep branch would be too early to go down that route. /semi-off-topic-wall-o-text

Reply:
Not sure if the circumstances are right to start this attack refactor. Reviewing has not even really started (PR was on hold for the last months) and the level of frustration is already alarming. Better drop the patches for now and let everyone take a break to recover if that is needed.

Reply:
@tovl: To be honest, your response hurts. The last release was a kick in the teeth for all of us, and I too am feeling burned out and still rather resentful about organizing most of my scarce free time over the summer around OpenRA. I invested more time than I was happy to give, at the expense of my personal and professional life, because it felt like I was the only person who really cared about the impact that the bugs in the previous release were having on our player community - and "just one more weekend" should have been enough for the stable release that never quite happened.

The situation sucks, but we all have to deal with it the best we can. I still feel that delaying this PR was the right choice and i'm sure that #17413 would have been a lot harder to resolve if this, #16739, and five months worth of followup work had been merged to bleed before somebody found it.

The overall direction of the attack refactor looks good, but its not fair or reasonable to try and tackle this in the same way as the activity rework. That was an awesome technical achievement, but it destroyed at least mine and @reaperrr's motivation, and I am still struggling to find any motivation to get back into reviewing. A large part of the problem was that the initial PRs were too ambitious, and broke too many unexpected assumptions that took too much work to identify after the fact and fix. I remember you agreeing on this point a few months ago.

The way to avoid this is, having worked out the broad shape of the refactoring, to take the extra time and extra PRs to rework the surrounding code and fix assumptions ahead of time. This focuses review and testing efforts, and makes it a lot easier to spot and fix any unexpected regressions before they become monstrous. Look at any of the big features I've implemented over the last few years to see that I hold myself to this standard as much (probably more) than anyone else.

Given this, I feel that my request about the `FacingTolerance` is completely reasonable. I would bet that the behaviour differences there would otherwise become a problem before the end of this, and that it will be a nightmare to fix if we wait until all of our combat units share 75% of their logic before doing so.

If you still think I'm being unreasonable then I agree that its probably for the best that we close this for now (but definitely park it on the Future milestone) until our collective spirits have raised enough to tackle it again.